Update custom UDF location guidance for WinUI3 unpackaged (#3845)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/concepts/user-data-folder.md

WebView2 checks for that runtime, and creates the UDF in that writeable location.

<!-- ------------------------------ -->
#### Packaging

A WinUI 2 (UWP) app is self-contained regarding dependencies and for deployment (DLLs), although it's not packaged into a single file.

A WinUI 2 (UWP) app is per-user, and has Write access under the installed location.

##### [WinUI 3](#tab/winui3)

On this platform, the default UDF location is the `ApplicationData\LocalFolder` subfolder in the package's folder.

<!-- ------------------------------ -->
#### Should you use default or custom UDF location?

<!-- todo: review -->

> On this platform, use the default UDF location if distributing as packaged.  If distributing as unpackaged, in scenarios where the install directory is protected, you must specify a custom UDF location.

Reviewer: review this paragraph 1 from Elijah
@david-risney 

https://review.learn.microsoft.com/microsoft-edge/webview2/concepts/user-data-folder?branch=pr-en-us-3845&tabs=winui3#the-default-udf-location

Reviewer 2: This sounds accurate

For example code, see the .NET (WPF & WinForms)-appropriate, .cpp or .cs file, near [WebView2Samples repo > WebView2WpfBrowser](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/SampleApps/WebView2WpfBrowser).

<!-- neither specific to custom nor default -->
**Where browser data gets stored within the UDF:**

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

##### [WinUI 2 (UWP)](#tab/uwp)

On this platform, use the default UDF location.

**Where browser data gets stored within the UDF:**

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

**Example code:**

For example code, see line 161 in [/webview2_sample_uwp/Pages/SettingsPage.xaml.cs](https://github.com/MicrosoftEdge/WebView2Samples/blob/main/SampleApps/webview2_sample_uwp/Pages/SettingsPage.xaml.cs#L161).  This sample app doesn't set a custom user data folder.  This code shows what the user data folder turned out to be.

##### [WinUI 3](#tab/winui3)

> On this platform, if distributing unpackaged to a protected install directory, you must specify a custom UDF location.

Reviewer: i added "as", consistent w/ line 188.
```suggestion
On this platform, if distributing as unpackaged to a protected install directory, you must specify a custom UDF location.
```
seems like a word is missing.  "distributing unpackaged" what?  Distributing an unpackaged foo?

but this pattern is used elsewhere too.

line 188 in WinUI 3 tab:
```
On this platform, use the default UDF location if distributing as packaged.  If distributing as unpackaged, in scenarios where the install directory is protected, you must specify a custom UDF location.
```

@david-risney @elibroftw

For example code, see the .NET (WPF & WinForms)-appropriate, .cpp or .cs file, near [WebView2Samples repo > WebView2WpfBrowser](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/SampleApps/WebView2WpfBrowser).

<!-- neither specific to custom nor default -->
**Where browser data gets stored within the UDF:**

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

##### [WinUI 2 (UWP)](#tab/uwp)

On this platform, use the default UDF location.

**Where browser data gets stored within the UDF:**

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

**Example code:**

For example code, see the WinUI 2 (UWP) `.cs` file, at [WebView2Samples repo > webview2_sample_uwp](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/SampleApps/webview2_sample_uwp).
<!-- todo: what file name?
App.xaml.cs
Settings.cs
WebView2Extensions.cs
-->

##### [WinUI 3](#tab/winui3)

On this platform, if distributing unpackaged to a protected install directory, you must specify a custom UDF location.

**Example code:**

> ```csharp

Reviewer: review this code from Elijah

see custom code in Issue also
https://github.com/MicrosoftDocs/edge-developer/issues/2001

@david-risney 

https://review.learn.microsoft.com/microsoft-edge/webview2/concepts/user-data-folder?branch=pr-en-us-3845&tabs=winui3#specifying-a-custom-udf-location

Reviewer 2: The WebView2 aspects of this good seem correct. The indenting in the catch block is off compared to the try block. The error handling looks a little strange but the comment is trying to indicate that its intentionally ambiguous. Overall seems good

Reviewer: fixed indent

try {
    // The default user data folder is next to the executable; however, the install dir (that is, /Program Files/) is protected.
    var userDataFolder = Path.Combine(
      Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),

>       "YOURAPPNAME", "WebView2"); // Decide whether you want to keep this folder after uninstall, or delete the folder by using the uninstaller.

Reviewer: review this revised code comment
orig commit: 
`// Decide whether you want to leave this after uninstall or clean it up from the uninstaller`

@david-risney 
@elibroftw

Reviewer 2: If this is intentionally saying to the end developer reading this code that they need to also decide if they want to change their uninstall code to remove this folder we may want to make this language more explicit. Eg: `// You need to decide if you want to keep this folder during uninstall or you need to add uninstall code that will remove this folder`. This is a reasonable point to bring up and document.

Reviewer: ```csharp
try {
    // The default user data folder is next to the executable; however, the install
    // dir (that is, /Program Files/) is protected.
    var userDataFolder = Path.Combine(
      Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
      "YOURAPPNAME", "WebView2"); // You need to decide if you want to keep this 
      // folder during uninstall, or add uninstall code that removes this folder.
    Log.Info($"WebView2 user data folder: {userDataFolder}");
    var env = await CoreWebView2Environment.CreateWithOptionsAsync(
        null, userDataFolder, new CoreWebView2EnvironmentOptions());
    await WebView.EnsureCoreWebView2Async(env);
} catch (Exception ex) {
    Log.Error($"WebView.EnsureCoreWebView2Async failed: {ex.Message}");
    // Implementation of fallback error UI is not provided.
    NativeUiError = ex.ToString();
    return;
}
```

For example code, see the .NET (WPF & WinForms)-appropriate, .cpp or .cs file, near [WebView2Samples repo > WebView2WpfBrowser](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/SampleApps/WebView2WpfBrowser).

<!-- neither specific to custom nor default -->
**Where browser data gets stored within the UDF:**

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

##### [WinUI 2 (UWP)](#tab/uwp)

On this platform, use the default UDF location.

**Where browser data gets stored within the UDF:**

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

**Example code:**

> For example code, see the WinUI 2 (UWP) `.cs` file, at [WebView2Samples repo > webview2_sample_uwp](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/SampleApps/webview2_sample_uwp).

Reviewer: can we specify the file name?

@david-risney

Reviewer 2: Afaict this sample app does not set a custom user data folder. It does look like it can show you what the user data folder turned out to be: https://github.com/MicrosoftEdge/WebView2Samples/blob/7fa88fc6fa222ce3eb05f277b02d1e6261b4e171/SampleApps/webview2_sample_uwp/Pages/SettingsPage.xaml.cs#L161

Reviewer: ```
For example code, see line 161 in 
[/webview2_sample_uwp/Pages/SettingsPage.xaml.cs](https://github.com/MicrosoftEdge/WebView2Samples/blob/main/SampleApps/webview2_sample_uwp/Pages/SettingsPage.xaml.cs#L161).
This sample app doesn't set a custom user data folder.  This code shows 
what the user data folder turned out to be.
```

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

##### [WinUI 2 (UWP)](#tab/uwp)

On this platform, use the default UDF location.

**Where browser data gets stored within the UDF:**

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

**Example code:**

For example code, see the WinUI 2 (UWP) `.cs` file, at [WebView2Samples repo > webview2_sample_uwp](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/SampleApps/webview2_sample_uwp).
<!-- todo: what file name?
App.xaml.cs
Settings.cs
WebView2Extensions.cs
-->

##### [WinUI 3](#tab/winui3)

<!-- todo: review -->
On this platform, if distributing unpackaged to a protected install directory, you must specify a custom UDF location.

**Example code:**

<!-- todo: review -->
```csharp
try {
    // The default user data folder is next to the executable; however, the install dir (that is, /Program Files/) is protected.
    var userDataFolder = Path.Combine(
      Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),

>       "YOURAPPNAME", "WebView2"); // Decide whether you want to keep this dir after uninstall, or clean it up from the uninstaller.  TODO: clarify "clean it up", ie delete the dir?

Reviewer: ```suggestion
      "YOURAPPNAME", "WebView2"); // Decide whether you want to keep this dir after uninstall, or delete the dir by using the uninstaller.
```
does "leave this" (in the orig commit) mean keep this dir?

does "clean it up" (in the orig commit) mean del the dir?

does "from the installer" (in the orig commit) mean "by using the installer?"

orig commit:
<img width="869" height="99" alt="image" src="https://github.com/user-attachments/assets/d36e2572-e7aa-4fb6-8813-d0c764c54d9e" />

@david-risney 
@elibroftw

Reviewer: ```
"YOURAPPNAME", "WebView2"); // Decide whether you want to keep this folder after uninstall, or delete the folder by using the uninstaller.
```

<!-- neither specific to custom nor default -->
<!-- ------------------------------ -->
#### Where browser data gets stored within the UDF

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

##### [WinUI 2 (UWP)](#tab/uwp)

On this platform, use the default UDF location.

<!-- ------------------------------ -->
#### Where browser data gets stored within the UDF

After creation of the session and UDF, browser data from your WebView2 control is stored in a subfolder of `userDataFolder`.

<!-- ------------------------------ -->
#### Example code

For example code, see the WinUI 2 (UWP) `.cs` file, at [WebView2Samples repo > webview2_sample_uwp](https://github.com/MicrosoftEdge/WebView2Samples/tree/main/SampleApps/webview2_sample_uwp).

##### [WinUI 3](#tab/winui3)

<!-- todo: review -->

> On this platform, if distributing unpackaged to a protected install directory, you must specify a custom UDF location.

Reviewer: review this paragraph 2 from Elijah
@david-risney 

https://review.learn.microsoft.com/microsoft-edge/webview2/concepts/user-data-folder?branch=pr-en-us-3845&tabs=winui3#specifying-a-custom-udf-location

Reviewer 2: Yes this sounds accurate

<!-- ------------------------------ -->
#### Example code

> <!-- todo: review -->

Reviewer: review this code from Elijah

see custom code in Issue also
https://github.com/MicrosoftDocs/edge-developer/issues/2001

@david-risney 

https://review.learn.microsoft.com/microsoft-edge/webview2/concepts/user-data-folder?branch=pr-en-us-3845&tabs=winui3#specifying-a-custom-udf-location

Reviewer: outdated
[fresh](https://github.com/MicrosoftDocs/edge-developer/pull/3845/files#r3493337660)

For examples of reading the `UserDataFolder` property, see the Win32 samples in the [WebView2Samples repo](https://github.com/MicrosoftEdge/WebView2Samples).

##### [.NET (WPF/WinForms)](#tab/dotnet)

<!-- Use the .NET [CoreWebView2Environment.UserDataFolder Property](/dotnet/api/microsoft.web.webview2.core.corewebview2environment.userdatafolder). -->

<!-- dev: add example code to https://learn.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2environment.userdatafolder showing how to read the `UserDataFolder` property, copy that from the below code block: -->

**Example code:**

<!-- ```csharp -->

> <!-- todo: AB#? // ADO work item "[wv2] Update .NET (WPF/WinForms) sample to add code to retrieve UDF location" - then copy lines to here

Reviewer: AB#?   [wv2] Update .NET (WPF/WinForms) sample to add code to retrieve UDF location

@captainbrosset

Reviewer: `<!-- todo: when AB#38572893 ([wv2] Update .NET (WPF/WinForms) sample to add code to retrieve UDF location) is fixed, copy code to here, add "additional" below`

<!-- ``` -->

For examples of reading the `UserDataFolder` property, see the .NET (WPF/WinForms) samples in the [WebView2Samples repo](https://github.com/MicrosoftEdge/WebView2Samples).

##### [ClickOnce](#tab/clickonce)

Use the .NET [CoreWebView2Environment.UserDataFolder Property](/dotnet/api/microsoft.web.webview2.core.corewebview2environment.userdatafolder).

**Example code:**

<!-- ```csharp -->

> <!-- todo: AB#? // ADO work item "[wv2] Update ClickOnce sample to add code to retrieve UDF location" - then copy lines to here

Reviewer: AB#?   [wv2] Update ClickOnce sample to add code to retrieve UDF location

@captainbrosset

Reviewer: `<!-- todo: when AB#38572793 ([wv2] Update ClickOnce sample to add code to retrieve UDF location) is fixed, copy code to here, add "additional" below`
